File: 이코테 복습(python, c++, java)/다이나믹 프로그래밍/c++/금광.cpp
#include<iostream>
using namespace std;

int testCase, n, m;
int arr[20][20];
int dp[20][20];

int main(void) {
	cin >> testCase;
	for (int tc = 0; tc < testCase; tc++) {
		cin >> n >> m;
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				cin >> arr[i][j];
			}
		}

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				dp[i][j] = arr[i][j];
			}
		}

		for (int j = 1; j < m; j++) {
			for (int i = 0; i < n; i++) {
				int leftUp, leftDown, left;

				if (i == 0) {
					leftUp = 0;
				}
				else {
					leftUp = dp[i - 1][j - 1];
				}
				if (i == n - 1) {
					leftDown = 0;
				}
				else {
					leftDown = dp[i + 1][j - 1];
				}

				left = dp[i][j - 1];
				dp[i][j] = dp[i][j] + max(leftUp, max(left, leftDown));
			}
		}

		int result = 0;
		for (int i = 0; i < n; i++) {
			result = max(result, dp[i][m - 1]);
		}
		cout << result << endl;
	}

}

File: 이코테 복습(python, c++, java)/다이나믹 프로그래밍/python/금광.py
T = int(input())
for _ in range(T):
    N, M = map(int, input().split())
    data = list(map(int, input().split()))
    dp = []
    index = 0
    for _ in range(N):
        dp.append(data[index:index + M])
        index += M

    for j in range(1, M):
        for i in range(N):
            if i == 0:
                left_up = 0
            else:
                left_up = dp[i - 1][j - 1]
            if i == N - 1:
                left_down = 0
            else:
                left_down = dp[i + 1][j - 1]

            left = dp[i][j - 1]
            dp[i][j] = dp[i][j] + max(left, left_up, left_down)
    result = 0
    for i in range(N):
        result = max(result, dp[i][M - 1])
    print(result)